artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/broker/AMQPMessage.java

```diff
@@ -709,30 +732,41 @@ public final void sendBuffer(ByteBuf buffer, int deliveryCount) {
     *
     * @return a Netty ByteBuf containing the encoded bytes of this Message instance.
     */
-   public ReadableBuffer getSendBuffer(int deliveryCount) {
+   public ReadableBuffer getSendBuffer(int deliveryCount, MessageReference reference) {
       ensureDataIsValid();
 
+      DeliveryAnnotations deliveryAnnotations;
+
+      if (reference != null && reference.getProtocolData() != null && reference.getProtocolData() instanceof DeliveryAnnotations) {
+         deliveryAnnotations = (DeliveryAnnotations) reference.getProtocolData();
+      } else {
+         deliveryAnnotations = null;
+      }
+
       if (deliveryCount > 1) {
-         return createCopyWithNewDeliveryCount(deliveryCount);
+         return createCopyWithNewDeliveryCount(deliveryCount, deliveryAnnotations);
       } else if (deliveryAnnotationsPosition != VALUE_NOT_PRESENT
+         || deliveryAnnotations != null
          || (deliveryAnnotationsForSendBuffer != null && !deliveryAnnotationsForSendBuffer.getValue().isEmpty())) {
-         return createCopyWithSkippedOrExplicitlySetDeliveryAnnotations();
+         return createCopyWithSkippedOrExplicitlySetDeliveryAnnotations(deliveryAnnotations);
```

Comment:
@gemmellr this is to allow EnMaas hack to still be useful. it's the old way they used here.. and I don't want to break their code.

Reply:
@clebertsuconic  can we expand on what this "hack" is please.

Reply:
@michaelpearce-gain  Project EnMaas is using extension to some of our components, and adding deliveryAnnotationsForSendBuffer on the message, as they needed to parse delivery annotation on their components. 

However this is only valid for a message... when the actual data we need for annotations comes from the reference.

So, I added a way to parse delivery annotations from messageReference, by translating the reference.protocolData.. while I am still preserving deliveryAnnotationsForSendBuffer...


deliveryAnnotationsForSendBuffer is sort of deprecated now... we will have to remove it at a later point.

Reply:
Im concerned we are dealing with third party projects extensions, they should be working with our apis, anything else, they should not be expecting compatibility....

Reply:
@michaelpear-gain: there are tests dealing with that.. at this point I consider this part of the API. although the tests were added by EnMaas guys themselves.

I'm keeping the old API while added a new better way to deal with it.

Reply:
I dont consider what was there a hack, at all, looking at the original commit its addition seems fairly consistent with how the message was being used at the time

I think the additions here also have utility.....but I think the combined impl with both approaches in play is rather incomprehensible. I also dont see an immediate need for the change.

I also raised another comment that I actually wondered (independently of this) whether a message (rather than delivery) annotation might even be more appropriate for the thing that seems to have sparked all this, given the way it is seemingly used.

Reply:
I can't use message-annotations, for this, as I would need to re-encode the message.

Notice that I am not re-encoding the message or copying it to send it over the replica... So, I don't have any other option other than the delivery annotation, which actually only makes sense through the replica wire.. (Hence delivery annotation).

Reply:
Another reason I can't use message-annotations is the message annotation would eventually be sent to the client... which I don't want to..


I really want the annotation to only exist during delivery.